File: xaiop-sdk/python/src/xaiop/control/resume_log.py
"""Outbound phase wire log for producer-side resume."""

from __future__ import annotations

from typing import Any

from .control import stamp_wire_with_log_seq


class XaiopResumeLogError(Exception):
    def __init__(
        self,
        message: str,
        *,
        code: str = "RESUME_LOG",
        seq: int | None = None,
    ) -> None:
        super().__init__(message)
        self.code = code
        self.seq = seq


class ResumeWireLog:
    def __init__(self) -> None:
        self._entries: list[dict[str, Any]] = []

    @property
    def size(self) -> int:
        return len(self._entries)

    @property
    def highest_seq(self) -> int:
        if not self._entries:
            return 0
        return self._entries[-1]["seq"]

    def record(self, entry: dict[str, Any]) -> ResumeWireLog:
        seq = entry.get("seq")
        if not isinstance(seq, int) or seq < 1:
            raise TypeError("ResumeWireLog.record requires seq >= 1")
        if not isinstance(entry.get("wire"), str):
            raise TypeError("ResumeWireLog.record requires wire string")
        last = self.highest_seq
        if seq <= last:
            raise XaiopResumeLogError(
                f"ResumeWireLog seq must be strictly increasing "
                f"(got {seq}, last {last})",
                code="RESUME_LOG_SEQ",
                seq=seq,
            )
        self._entries.append(
            {
                "seq": seq,
                "wire": entry["wire"],
                "committed": entry.get("committed"),
            }
        )
        return self

    def wires_after(self, from_seq: int) -> str:
        return self._join_after(from_seq, True)

    def wires_after_raw(self, from_seq: int) -> str:
        return self._join_after(from_seq, False)

    def _join_after(self, from_seq: int, stamp: bool) -> str:
        n = int(from_seq)
        if n < 0:
            raise TypeError("wires_after requires non-negative integer fromSeq")
        out = ""
        for e in self._entries:
            if e["seq"] > n:
                out += (
                    stamp_wire_with_log_seq(e["seq"], e["wire"])
                    if stamp
                    else e["wire"]
                )
        return out

    def clear(self) -> ResumeWireLog:
        self._entries.clear()
        return self

File: xaiop-sdk/go/xaiop/control/resume_log.go
package control

import "fmt"

// ResumeLogError is a ResumeWireLog sequencing failure.
type ResumeLogError struct {
	Message string
	Code    string
	Seq     int
}

func (e *ResumeLogError) Error() string {
	if e == nil {
		return "resume log error"
	}
	return e.Message
}

// ResumeEntry is one outbound phase record.
type ResumeEntry struct {
	Seq       int
	Wire      string
	Committed any
}

// ResumeWireLog stores producer-side phase wire for resume.
type ResumeWireLog struct {
	entries []ResumeEntry
}

// NewResumeWireLog creates an empty log.
func NewResumeWireLog() *ResumeWireLog {
	return &ResumeWireLog{}
}

// Size returns entry count.
func (l *ResumeWireLog) Size() int { return len(l.entries) }

// HighestSeq returns the last recorded seq (0 if empty).
func (l *ResumeWireLog) HighestSeq() int {
	if len(l.entries) == 0 {
		return 0
	}
	return l.entries[len(l.entries)-1].Seq
}

// Record appends a strictly-increasing seq entry.
func (l *ResumeWireLog) Record(entry ResumeEntry) error {
	if entry.Seq < 1 {
		return fmt.Errorf("ResumeWireLog.record requires seq >= 1")
	}
	if entry.Wire == "" && entry.Seq > 0 {
		// wire may be empty string legitimately; only reject non-string via type — Go is typed
	}
	last := l.HighestSeq()
	if entry.Seq <= last {
		return &ResumeLogError{
			Message: fmt.Sprintf("ResumeWireLog seq must be strictly increasing (got %d, last %d)", entry.Seq, last),
			Code:    "RESUME_LOG_SEQ",
			Seq:     entry.Seq,
		}
	}
	l.entries = append(l.entries, entry)
	return nil
}

// WiresAfter returns stamped wires with seq > fromSeq.
func (l *ResumeWireLog) WiresAfter(fromSeq int) (string, error) {
	return l.joinAfter(fromSeq, true)
}

// WiresAfterRaw returns raw wires with seq > fromSeq (no stamp).
func (l *ResumeWireLog) WiresAfterRaw(fromSeq int) (string, error) {
	return l.joinAfter(fromSeq, false)
}

func (l *ResumeWireLog) joinAfter(fromSeq int, stamp bool) (string, error) {
	if fromSeq < 0 {
		return "", fmt.Errorf("wires_after requires non-negative integer fromSeq")
	}
	out := ""
	for _, e := range l.entries {
		if e.Seq <= fromSeq {
			continue
		}
		if stamp {
			s, err := StampWireWithLogSeq(e.Seq, e.Wire)
			if err != nil {
				return "", err
			}
			out += s
		} else {
			out += e.Wire
		}
	}
	return out, nil
}

// Clear empties the log.
func (l *ResumeWireLog) Clear() *ResumeWireLog {
	l.entries = nil
	return l
}
